### PR TITLE
stdlib: remove `Domain.at_each_spawn`

### LINKED ISSUE #11434: NULL pointer dereferencing and use-after-free in systhreads due to delayed init
This is a follow-up to #11403, which noticed that the domain threads initialization code runs from OCaml code. Thus there is an interval during which arbitrary OCaml code can run before initialization (e.g. GC, async callbacks and at_each_spawn callbacks). This can cause NULL pointer dereferencing and use-after-free inside `caml_thread_self` and `caml_enter_blocking_section`. Following the discussion at #11403, it appears that the best solution is:

- [ ] Replace the use of `Domain.at_each_spawn` to call `caml_thread_initialize_domain` with a C hook `caml_domain_at_each_spawn` similar to `caml_domain_stop_hook` (internal, non-atomic).
- [ ] Optionally remove the API `Domain.at_each_spawn`, which is a bad API for other reasons discussed somewhere else.

Targets 5.0.

### PR DESCRIPTION
**Pull request by Octachron**
This PR proposes to remove the function `Domain.at_each_spawn`. 

One major issue with `Domain.at_each_spawn` is that it is impossible to predict in which context the registered function will end up being called. 

A good illustration of the issue at hand is the use of `Domain.at_each_spawn` in the threads library in order to initialize the thread machinery. This approach breaks down if the order of the initialization is incorrect as described in #11434.  Consider for instance, a module `Early_hook` registering a function with `Domain.at_each_spawn`
```ocaml
(* early_hook.ml *)
let f = ref ignore
let () = Domain.at_each_spawn (fun () -> !f ())
```
and the following main function:
```ocaml
(* main.ml *)
let print_self () =
  Printf.printf "self=%d\n" (Thread.id (Thread.self ()))
let () =
  Early_hook.f := print_self;
  for i = 0 to 50 do
    Domain.join (Domain.spawn ignore)
  done
```
With this setup, any program compiled with `early_hook` linked before the `systhreads` library crashes because the function registered by `early_hook` calls `Thread.self` before the thread machinery initialization.

The first commit in this PR fixes this issue by adding a new C hook `caml_domain_initialize_hook` for the systhreads library which is run during domain initialization just before the registration with the stop-the-world mechanism. (I am not sure of the right location for the hook during initialization). The second commit adds the example described above as an ephemeral test.

The two last commits removes the use of `Domain.at_each_spawn` in `Format` by moving the registration of the `Domain.at_exit` functions to the initialization of the by-domain stdout and stder formatter.

Finally, the last commit completely remove `Domain.at_each_spawn` function since the function seems more harmful than helpful in its current state. It is thus probably better in term of forward compatibility to remove it.

Close #11434

### CODE REVIEW

File: runtime/domain.c [hunk 1]
```diff
@@ -688,7 +688,7 @@ static void domain_create(uintnat initial_minor_heap_wsize) {
 #endif
 
   caml_reset_young_limit(domain_state);
-
+  caml_domain_initialize_hook();
```
Comment (gadmm) [inline runtime/domain.c]: This is called while holding the `all_domains` lock. It is better to wait that the lock is released. I would call `caml_domain_initialize_hook` right before the call to `caml_callback` inside `domain_thread_func`.

**Review by gadmm — APPROVED** [review]
Overall this looks good to me apart from the location where the hook is called, but I had a fairly quick look. Would @Engil or @gasche be willing to give a second opinion?

**Review by kayceesrk — APPROVED** [review]
The change looks good to me. Removing `Domain.at_each_spawn` is a good move.

File: runtime/domain.c (hunk 1 above)
Comment (Octachron) [reply runtime/domain.c]: That seems better indeed.

**Comment by Octachron** [comment]
I have cleaned the history and I am planning to merge tomorrow in order to move forward the release cycle.

**Pull request merged**